examples/chart/teleport-kube-agent/templates/hook.yaml

```diff
@@ -0,0 +1,93 @@
+{{- $deployment := (lookup "apps/v1" "Deployment"  .Release.Namespace .Release.Name ) -}}
+{{- if $deployment }}
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Release.Name }}-hook
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": post-upgrade
+    "helm.sh/hook-weight": "-4"
+    "helm.sh/hook-delete-policy": hook-succeeded
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ .Release.Name }}-hook
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": post-upgrade
+    "helm.sh/hook-weight": "-3"
+    "helm.sh/hook-delete-policy": hook-succeeded
+rules:
+  - apiGroups: ["apps"]
+    resources: ["statefulsets"]
+    resourceNames: ["{{ .Release.Name }}"]
+    verbs: ["get", "watch", "list"]
+  - apiGroups: [""]
+    resources: ["pods",]
+    verbs: ["get", "watch"]
+  - apiGroups: ["apps"]
+    resources: ["deployments",]
+    resourceNames: ["{{ .Release.Name }}"]
+    verbs: ["get", "delete", "list"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ .Release.Name }}-hook
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": post-upgrade
+    "helm.sh/hook-weight": "-2"
+    "helm.sh/hook-delete-policy": hook-succeeded
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ .Release.Name }}-hook
+subjects:
+- kind: ServiceAccount
+  name: {{ .Release.Name }}-hook
+  namespace: {{ .Release.Namespace }}
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: "{{ .Release.Name }}"-hook
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": post-upgrade
+    "helm.sh/hook-weight": "-1"
+    "helm.sh/hook-delete-policy": hook-succeeded
+spec:
+  template:
+    metadata:
+      name: "{{ .Release.Name }}"-hook
+    spec:
+      serviceAccountName: {{ .Release.Name }}-hook
+      restartPolicy: OnFailure
+      containers:
+      - name: post-install-job
+        image: "mirror.gcr.io/library/alpine"
+        command: 
+        - sh
+        - "-c"
+        - |
+            /bin/sh <<'EOF'
+              set -eu -o pipefail
+              # download curl
+              apk add curl
+              # downlad kubectl
+              curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+              curl -LO "https://dl.k8s.io/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl.sha256"
+              echo "$(cat kubectl.sha256)  kubectl" | sha256sum -c
+              if [ $? -ne 0 ]; then
+                  exit $?
+              fi
+              chmod +x kubectl
+              # wait until statefulset is ready
+              ./kubectl rollout status --watch --timeout=600s statefulset/{{ .Release.Name }}
+              # delete deployment
+              ./kubectl delete deployment/{{ .Release.Name }}
```

Comment:
Should these commands specify release namespace?

Reply:
Not necessary. By default `kubectl` loads the credentials and namespace from `/var/run/secrets/kubernetes.io/serviceaccount`